mathjax3-ts/output/chtml.ts

```diff
@@ -53,258 +38,64 @@ type MetricDomMap<N> = Map<N, N>;
-     */
-    protected getMetricMap(html: MathDocument<N, T, D>) {
-        const adaptor = this.adaptor;
-        const domMap = new Map() as MetricDomMap<N>;
-        //
-        // Add the test elements all at once (so only one reflow)
-        //
-        for (const math of html.math) {
-            const parent = adaptor.parent(math.start.node);
-            if (!domMap.has(parent)) {
-                domMap.set(parent, this.getTestElement(parent));
-            }
-        }
-        //
-        // Measure the metrics for all the mapped elements
-        //
-        const map = new Map() as MetricMap<N>;
-        for (const node of domMap.keys()) {
-            map.set(node, this.measureMetrics(domMap.get(node)));
-        }
-        //
-        // Remove the test elements
-        //
-        for (const node of domMap.values()) {
-            adaptor.remove(node);
-        }
-        return map;
-    }
-
-    /**
-     * @param {N} node    The container to add the test elements to
-     * @return {N}        The test elements that were added
-     */
-    protected getTestElement(node: N) {
-        const adaptor = this.adaptor;
-        if (!this.testNodes) {
-            this.testNodes = this.html('mjx-test', {style: {
-                display:           'inline-block',
-                'font-style':      'normal',
-                'font-weight':     'normal',
-                'font-size':       '100%',
-                'font-size-adjust':'none',
-                'text-indent':     0,
-                'text-transform':  'none',
-                'letter-spacing':  'normal',
-                'word-spacing':    'normal',
-                overflow:          'hidden',
-                height:            '1px'
-            }}, [
-                this.html('mjx-ex-box', {style: {
-                    position: 'absolute',
-                    overflow: 'hidden',
-                    width: '1px', height: '60ex'
-                }})
-            ]);
-        }
-        return adaptor.append(node, adaptor.clone(this.testNodes)) as N;
-    }
-
-    /**
-     * @param {N} node    The test node to measure
-     * @return {Metrics}  The metric data for the given node
-     */
-    protected measureMetrics(node: N) {
-        const adaptor = this.adaptor;
-        const em = adaptor.fontSize(node);
-        const ex = (adaptor.nodeSize(adaptor.firstChild(node) as N)[1] / 60) || (em * this.options.exFactor);
-        const containerWidth = 80 * em; // for now (should be measured in the DOM)
-        const scale = ex / this.font.params.x_height / em;
-        const lineWidth = 1000000;      // no linebreaking (otherwise would be a percentage of cwidth)
-        return {em, ex, containerWidth, lineWidth, scale} as Metrics;
+    public styleSheet(html: MathDocument<N, T, D>) {
+        const sheet = super.styleSheet(html);
+        this.adaptor.setAttribute(sheet, 'id', 'CHTML-styles');
```

Comment:
The `updateDocument()` method that calls this will remove any earlier item with `id="CHTML-styles"` before inserting this.  That is so that if a second round of typesetting is done, the styles will be updated.  The idea is that the CSS might be tied to the content (e.g., only the characters that are actually used will be included), so that allows for updating the stylesheet.

We may want to work out a more general approach to that in the future.